src/datumaro/plugins/data_formats/coco/page_mapper.py

```diff
@@ -261,31 +321,42 @@ def update(braket, brace, c, cursor):
         else:
             brace.buffer += c
 
-    def __iter__(self) -> Iterator[Tuple[Dict, Dict]]:
-        reader = FileReaderWithCache(self.path)
-
+    def __iter__(self) -> Iterator[Tuple[Dict, List[Dict]]]:
         for item_page in self.item_page_map.values():
-            data = reader.read(item_page.offset, item_page.size)
-            item_dict = parse_json(data)
-            anns_dict = []
-
-            curr_pt = item_page.ann_last_pt
-
-            to_read = []
-            while curr_pt >= 0:
-                ann_offset = self.ann_page_map.offsets[curr_pt]
-                ann_size = self.ann_page_map.sizes[curr_pt]
-                curr_pt = self.ann_page_map.prev_pts[curr_pt]
-                to_read += [(ann_offset, ann_size)]
-
-            for ann_offset, ann_size in reversed(to_read):
-                data = reader.read(ann_offset, ann_size)
-                ann_dict = parse_json(data)
-                anns_dict.append(ann_dict)
-
-            yield item_dict, anns_dict
-
-        reader.close()
+            yield self._gen_item_dict(item_page), self._gen_anns_dict(item_page)
+
+    def get_item_dict(self, item_key: int) -> Optional[Dict]:
+        item_page = self.item_page_map.get(item_key)
+        if item_page is None:
+            return None
+        return self._gen_item_dict(item_page)
+
+    def get_anns_dict(self, item_key: int) -> Optional[List[Dict]]:
+        item_page = self.item_page_map.get(item_key)
+        if item_page is None:
+            return None
+        return self._gen_anns_dict(item_page)
+
+    def _gen_item_dict(self, item_page: ItemPage) -> Dict:
+        data = self.item_reader.read(item_page.offset, item_page.size)
+        return parse_json(data)
+
+    def _gen_anns_dict(self, item_page: ItemPage) -> List[Dict]:
+        curr_pt = item_page.ann_last_pt
```

Comment:
Simple question, why do you read from last?

Reply:
This is the head pointer of the linked list. As you know, we have a linked list of annotation pages. To iterate over the linked list, the first step is that we have to go to the head of the linked list.

Reply:
understood. thanks.